File: Data Structures & Algorithms/sliding-window-maximum/submission-2.py
class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        n = len(nums)
        lmax = [0]*n
        rmax = [0]*n
        currmax = nums[0]
        for i in range(n):
            if i%k == 0:
                currmax = nums[i]
            currmax = max(currmax,nums[i])
            lmax[i] = currmax
        
        currmax = nums[n-1]
        for i in range(n-1,-1,-1):
            if i%k == 0:
                currmax = nums[i]
            currmax = max(currmax,nums[i])
            rmax[i] = currmax

        ans = []
        for i in range(k-1,n):
            ans.append(max(rmax[int(i-k+1)],lmax[i]))

        return ans

File: Data Structures & Algorithms/jump-game-ii/submission-2.cpp
class Solution {
public:
    int jump(vector<int>& nums) {
        int n=nums.size();
        if(n==1) return 0;
        vector<int> dp(n+1,0);
        int farthest=0,currrange=nums[0],ans=1;

        for(int i=0;i<n-1;i++){
            farthest = max(farthest,i+nums[i]);
            if(i==currrange){
                ans++;
                currrange=farthest;
            }
        }
        return ans;
    }
};
